File: contents/huffman_encoding/code/java/huffman.java
import java.util.*;

class TreeNode {
    String letter = "";
    int frequency = 0;
    TreeNode left = null, right = null;

    public TreeNode(String letter, int frequency) {
        this.letter = letter;
        this.frequency = frequency;
    }

    public TreeNode(int frequency, TreeNode left, TreeNode right) {
        this.frequency = frequency;
        this.left = left;
        this.right = right;
    }
}

class HuffmanTree {
    private Map<String, Integer> frequencyMap = new HashMap<>();
    private Map<String, String> codeBook = new HashMap<>(), reverseCodeBook = new HashMap<>();
    private TreeNode root;
    private String stringToEncode;

    public HuffmanTree(String stringToEncode) {
        this.stringToEncode = stringToEncode;
    }

    public void createTree() {
        for (int i = 0; i < stringToEncode.length(); i++) {
            String key = Character.toString(stringToEncode.charAt(i));
            if (!frequencyMap.containsKey(key)) {
                frequencyMap.put(key, 1);
            } else {
                int frequency = frequencyMap.get(key) + 1;
                frequencyMap.replace(key, frequency);
            }
        }
        Queue<TreeNode> priorityQueue = new PriorityQueue<>(Comparator.comparingInt(o -> o.frequency));
        for (Map.Entry<String, Integer> m : frequencyMap.entrySet()) {
            priorityQueue.add(new TreeNode(m.getKey(), m.getValue()));
        }
        while (priorityQueue.size() > 1) {
            TreeNode left = priorityQueue.remove();
            TreeNode right = priorityQueue.remove();
            priorityQueue.add(new TreeNode(left.frequency + right.frequency, left, right));
        }
        root = priorityQueue.remove();
    }

    private void traverse(TreeNode node, StringBuilder code) {
        if (node.left == null && node.right == null) {
            codeBook.put(node.letter, code.toString());
        }
        if (node.left != null) {
            traverse(node.left, code.append(0));
            code.deleteCharAt(code.length() - 1);
        }
        if (node.right != null) {
            traverse(node.right, code.append(1));
            code.deleteCharAt(code.length() - 1);
        }
    }

    public void printCodeBook() {
        System.out.println("Code Book");
        for (Map.Entry<String, String> m : codeBook.entrySet()) {
            System.out.println(m.getKey() + "\t" + m.getValue());
        }
        System.out.println();
    }

    private void CodeBookReverse() {
        for (Map.Entry<String, String> m : codeBook.entrySet()) {
            reverseCodeBook.put(m.getValue(), m.getKey());
        }
    }

    public String encode() {
        traverse(root, new StringBuilder());
        StringBuilder encode = new StringBuilder();
        for (int i = 0; i < stringToEncode.length(); i++) {
            String k = Character.toString(stringToEncode.charAt(i));
            encode.append(codeBook.get(k));
        }
        printCodeBook();
        return encode.toString();
    }

    public String decode(String encoded) {
        StringBuilder decoded = new StringBuilder(), key = new StringBuilder();
        CodeBookReverse();
        for (int i = 0; i < encoded.length(); i++) {
            key = key.append(encoded.charAt(i));
            if (reverseCodeBook.containsKey(key.toString())) {
                decoded.append(reverseCodeBook.get(key.toString()));
                key = new StringBuilder();
            }
        }
        return decoded.toString();
    }
}

class Huffman {
    public static void main(String[] args) {
        HuffmanTree huffmanTree = new HuffmanTree("bibbity_bobbity");
        huffmanTree.createTree();
        String encoded = huffmanTree.encode();
        System.out.println("Encoded String: " + encoded);
        System.out.println("Decoded String: " + huffmanTree.decode(encoded));
    }
}

File: contents/huffman_encoding/code/kotlin/huffman.kt
import java.util.PriorityQueue

// Node type
sealed class Node(val frequency: Int): Comparable<Node> {
    // So can sort nodes by frequency
    override fun compareTo(other: Node) = frequency - other.frequency
}

class Leaf(freq: Int, val letter: Char): Node(freq)
class Branch(val left: Node, val right: Node): Node(left.frequency + right.frequency)


// Class to contain the decoder and encoder
class HuffmanTree (textSample: String) {
    private val root: Branch
    private val codeBook: MutableMap<Char, String>

    init {
        // Calculate frequencies
        val frequencyMap = textSample.groupingBy { it }.eachCount()

        // Populate the queue with leaves
        val priorityQueue = PriorityQueue<Node>(frequencyMap.map{(char,freq)->
            Leaf(freq, char)
        })

        // Join leaves
        while (priorityQueue.size > 1) {
            val left = priorityQueue.remove()
            val right = priorityQueue.remove()
            priorityQueue.add(Branch(left, right))
        }

        // The root node is the last remaining branch. If it's a leaf, we only had 1 unique letter
        root = priorityQueue.remove() as? Branch ?: error("No support for string of one unique letter")

        // Initialize the code book
        codeBook = mutableMapOf()
        populateCodeBook(root,"")
    }

    // Recursively populate a codebook with encodings from a node
    private fun populateCodeBook(node: Node, code: String) {
        when(node) {
            // Leaf node, set this char's code
            is Leaf -> codeBook[node.letter] = code

            // Branch node, recursively populate the children
            is Branch -> {
                populateCodeBook(node.left, code + "0")
                populateCodeBook(node.right, code + "1")
            }
        }
    }

    // Print the code book to stdout
    fun print() {
        for ((k,v) in codeBook.entries) {
            println("$k: $v")
        }
        println()
    }

    // Encode a string using this tree
    fun encode(source: String): String {
        val encoder = StringBuilder()
        for (char in source) {
            encoder.append(codeBook[char])
        }
        return encoder.toString()
    }


    // Decode an encoded string
    fun decode(encoded: String) = buildString {
        val endNode = encoded.fold(root){current, char->
            val next = when(char){
                '0' -> current.left
                '1' -> current.right
                else -> error("Found `$char` in the encoded string")
            }
            when(next){
                is Branch -> next
                is Leaf -> {
                    append(next.letter)
                    root
                }
            }
        }

        if(endNode != root)
            error("Unexpected end of string")
    }
}


fun main() {
    val sourceText = "bibbity_bobbity"
    // Create huffman tree
    val huffmanTree = HuffmanTree(sourceText)
    // Encode the text
    val encoded = huffmanTree.encode(sourceText)
    println("Encoded String: $encoded")
    println("Decoded String: ${huffmanTree.decode(encoded)}")

    println("CodeBook:")
    huffmanTree.print()
}
